src/transformStateWithClones.js

```diff
@@ -7,7 +7,39 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const newState = {};
+
+  Object.assign(newState, state);
+
+  const arrStateClone = [];
+
+  for (const action of actions) {
+    if (action.type === 'clear') {
+      Object.keys(newState).forEach(key => {
+        delete newState[key];
+      });
+
+      arrStateClone.push({ ...newState });
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+
+      arrStateClone.push({ ...newState });
+    }
+
+    if (action.type === 'removeProperties') {
+      const keyToRemove = action.keysToRemove;
+
+      for (const key of keyToRemove) {
+        delete newState[key];
+      }
+
+      arrStateClone.push({ ...newState });
+    }
```

Comment:
- it's better to use switch case statement instead of multiple if blocks 
- clear it's a magic words, so please move it and other magic words to the constants 
- `arrStateClone.push({ ...newState });` is duplicated, so make sense to move it after switch case block